chapter-08/recipe-07/cxx-example/src/evolution/evolution.cpp

```diff
@@ -6,21 +6,16 @@
 std::vector<int> evolve(const std::vector<int> row, const std::string rule_binary) {
   std::vector<int> result;
 
-  for (int i = 0; i < row.size(); i++) {
+  for (auto i = 0; i < row.size(); ++i) {
 
-    int left = i - 1;
-    int center = i;
-    int right = i + 1;
+    auto left = (i==0 ? row.size() : i) - 1;
+    auto center = i;
+    auto right = (i + 1) % row.size();
 
-    if (left < 0)
-      left += row.size();
-    if (right >= row.size())
-      right -= row.size();
-
-    int ancestors = 4 * row[left] + 2 * row[center] + 1 * row[right];
+    auto ancestors = 4 * row[left] + 2 * row[center] + 1 * row[right];
     ancestors = 7 - ancestors;
 
-    int new_state = std::stoi(rule_binary.substr(ancestors, 1));
+    auto new_state = std::stoi(rule_binary.substr(ancestors, 1));
```

Comment:
Mac build is choking on this line. Probably it needs `#include <cstdlib>`?

Reply:
May be but I don't see why?
What is the message?

This looks like a strange side effect of `auto`, because it should evaluate to return type of `std::stoi` which is `int` coming from `#include <string>` ??

Reply:
I shouldn't look at code too late at night. The error is in `parser.cpp` where you kept `int` instead of using `auto`:
```
cd /Users/travis/build/bast/cmake-recipes/chapter-08/recipe-07/cxx-example/build-2018-04-02-10-15-57/src/parser && /usr/bin/clang++   -I/Users/travis/build/bast/cmake-recipes/chapter-08/recipe-07/cxx-example/src/parser  -std=c++11 -o CMakeFiles/parser.dir/parser.cpp.o -c /Users/travis/build/bast/cmake-recipes/chapter-08/recipe-07/cxx-example/src/parser/parser.cpp
/Users/travis/build/bast/cmake-recipes/chapter-08/recipe-07/cxx-example/src/parser/parser.cpp:9:16: error: no member named 'stoi' in namespace 'std'; did you mean 'atoi'?
  int length = std::stoi(argv[1]);
               ^~~~~~~~~
               atoi
/usr/include/stdlib.h:132:6: note: 'atoi' declared here
int      atoi(const char *);
         ^
/Users/travis/build/bast/cmake-recipes/chapter-08/recipe-07/cxx-example/src/parser/parser.cpp:10:19: error: no member named 'stoi' in namespace 'std'; did you mean 'atoi'?
  int num_steps = std::stoi(argv[2]);
                  ^~~~~~~~~
                  atoi
/usr/include/stdlib.h:132:6: note: 'atoi' declared here
int      atoi(const char *);
         ^
/Users/travis/build/bast/cmake-recipes/chapter-08/recipe-07/cxx-example/src/parser/parser.cpp:11:22: error: no member named 'stoi' in namespace 'std'; did you mean 'atoi'?
  int rule_decimal = std::stoi(argv[3]);
                     ^~~~~~~~~
                     atoi
/usr/include/stdlib.h:132:6: note: 'atoi' declared here
int      atoi(const char *);
```